### DISCUSSION TITLE
Questions I asked as .NET interviewer

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#80] [General]
# Questions I asked as .NET interviewer

This blog post is a small collection of questions I asked as a technical interviewer when we had candidates for software developer positions. This might be helpful for you if you are preparing for a job interview.

https://steven-giesel.com/blogPost/d64172d0-6b72-4d60-9935-0159c72997e4

**Comment by wangengzheng** [comment]
good job!  I have carefully read it and think it is written very well.

**Comment by kyle69** [comment]
Hi!
Great approach to interviewing a candidate! Much more interesting than a Coding Game type test...

I have a little question, i don't see how: GetCustomers().ToList().Count() would produce two calls to the database?

The .ToList() will force the immediate evaluation of the query from the database which will "freeze" the items in a List<Customer> and the following .Count() call will not call the database because it will be done on the List<Customer>.

Thanks!

**Reply by linkdotnet** [reply]
Thanks for the kind words.

I guess I should have made that more clear. The case is not GetCustomers().ToList().Count() but calling GetCustomers().ToList() first and then GetCustomers().Count().

**Reply by kyle69** [reply]
ah okay, actually in this case I understand better ;)

**Reply by linkdotnet** [reply]
If I have access to my laptop, I will clarify that part better. Thanks for that, much appreciated

**Comment by dlavysh** [comment]
Thanks for the article. As always pleasure to read.
In Section "IEnumerable/IQuerable" i expected that you write something like:"
_IQueryable is suitable for querying data from out-memory (like remote database, service) collections. Whereas IEnumerable is good for In-Memory Collection query. While querying data from a database, IQueryable executes a "select query" on server-side with all filters_"
Maybe it should be obvious from your text, but still :)

**Reply by linkdotnet** [reply]
That is another good question to understand the difference between those two.

**Comment by LexaGV** [comment]
When man asks about design patters, he immediately downgrades in my eyes. DP means man cannot think on problem, but plays with many solutions of another problems, ZERO THINKING. All you need is to know kinds of DP, but applying 'em is strictly man's choice. In other words, it's the most stupid question on interview. Sounds like you test mathematician: hey, we have 2,3,4,5 numbers, to which equations they suit? - sounds silly as author.

**Reply by linkdotnet** [reply]
That is fine. Many frameworks like ASP.NET Core are using design patterns. 
In my opinion they have a "right to exist". That doesn't mean one should apply them blindly. But again, these are just my thoughts and opinions.

If someone doesn't like them - fine, but at least they should have a reasoning behind them. And as often, that means you need to understand something to know why you don't like it.

**Comment by LexaGV** [comment]
DI - just another buzzword, which is mostly liked by mediocres. DI is just one of billions solutions to IT problems, why you concentrate on this one?? Why you don't ask about bubble sort? Path finding? Balancing tree? ALL of 'em also solutions! :)  Level of your questions shows YOUR skills, not interviewed guy.

**Reply by linkdotnet** [reply]
Why should be "just a buzzword" - especially for quote: "mediocres"?

Many projects I worked with, especially in my former company, used IoC and used DI to achieve that.
I never stated that these are universal questions applicable to everyone - they are tailored to my former company.
Path finding, balanced binary tree and such were not important for the projects - if they would have been, yes I'd have asked such questions.

**Comment by dennyferra** [comment]
Just a nitpick, in section Inheritance vs. Composition, I believe the quote is "Favor composition over inheritance" (you have "Favorite")

**Reply by linkdotnet** [reply]
Thanks and indeed. Fixed it

**Comment by itamarep** [comment]
Some interviewers also like to ask questions on data structures (List, Queue, Stack, Tree, Linked list...)
